src/main/java/rx/internal/operators/OnSubscribeAmb.java

```diff
@@ -373,26 +374,46 @@ public void call() {
         }));
         subscriber.setProducer(new Producer() {
 
+            private final AtomicBoolean subscribedToSources = new AtomicBoolean(false);
+            
             @Override
             public void request(long n) {
                 if (selection.choice.get() != null) {
                     // propagate the request to that single Subscriber that won
                     selection.choice.get().requestMore(n);
-                } else {
+                } else if (subscribedToSources.compareAndSet(false, true)){
+                    //need to subscribe to all the sources
                     for (Observable<? extends T> source : sources) {
                         if (subscriber.isUnsubscribed()) {
-                            break;
+                            return;
                         }
                         AmbSubscriber<T> ambSubscriber = new AmbSubscriber<T>(n, subscriber, selection);
                         selection.ambSubscribers.add(ambSubscriber);
-                        // possible race condition in previous lines ... a choice may have been made so double check (instead of synchronizing)
+                        // check again if choice has been made so can stop subscribing
                         if (selection.choice.get() != null) {
                             // Already chose one, the rest can be skipped and we can clean up
                             selection.unsubscribeOthers(selection.choice.get());
-                            break;
+                            return;
                         }
                         source.unsafeSubscribe(ambSubscriber);
                     }
+                } else {
+                    //subscriptions already happened so propagate the request to all the  
+                    //amb subscribers
+                    for (AmbSubscriber<T> ambSubscriber: selection.ambSubscribers) {
```

Comment:
This may still run concurrently with the subscription loop and some subscribers won't get the newer requests during that time. The simplest fix is to subscribe all AmbSubscriber in the call() of the operator with 0 initial requests so all subsequent request() can target all or the winner without any additional race. The other way is to get into some complicated serialization logic.